common/authorization/interceptor.go

```diff
@@ -45,42 +50,79 @@ func (a *interceptor) Interceptor(
 	handler grpc.UnaryHandler,
 ) (interface{}, error) {
 
-	var namespace string
-	requestWithNamespace, ok := req.(requestWithNamespace)
-	if ok {
-		namespace = requestWithNamespace.GetNamespace()
-	}
+	var caller *Claims
 
-	apiName := info.FullMethod
-	index := strings.LastIndex(apiName, "/")
-	if index > -1 {
-		apiName = apiName[index+1:]
+	if a.claimMapper != nil {
+		var tlsSubject *pkix.Name
+		var authHeaders []string
+
+		if md, ok := metadata.FromIncomingContext(ctx); ok {
+			authHeaders = md["authorization"]
+		}
+		if p, ok := peer.FromContext(ctx); ok {
+			if tlsAuth, ok := p.AuthInfo.(credentials.TLSInfo); ok {
+				if len(tlsAuth.State.VerifiedChains) > 0 && len(tlsAuth.State.VerifiedChains[0]) > 0 {
+					tlsSubject = &tlsAuth.State.VerifiedChains[0][0].Subject
+				}
+			}
+		}
+		// Add auth into to ctx only if there's some auth info
+		if tlsSubject != nil || len(authHeaders) > 0 {
+			authInfo := AuthInfo{authHeaders[0], tlsSubject}
+			if a.authorizer != nil {
+				claims, err := a.claimMapper.GetClaims(authInfo)
+				if err != nil {
+					return nil, err
+				}
+				caller = claims
+				ctx = context.WithValue(ctx, "auth-claims", claims)
+			}
+		}
 	}
 
-	scope := a.getMetricsScope(metrics.NumAuthorizationScopes, namespace)
-	sw := scope.StartTimer(metrics.ServiceAuthorizationLatency)
-	defer sw.Stop()
+	if a.authorizer != nil {
+		var namespace string
+		requestWithNamespace, ok := req.(requestWithNamespace)
+		if ok {
+			namespace = requestWithNamespace.GetNamespace()
+		}
 
-	result, err := a.authorizer.Authorize(ctx, &Attributes{Namespace: namespace, APIName: apiName})
-	if err != nil {
-		scope.IncCounter(metrics.ServiceErrAuthorizeFailedCounter)
-		return nil, err
-	}
-	if result.Decision != DecisionAllow {
-		scope.IncCounter(metrics.ServiceErrUnauthorizedCounter)
-		return nil, errUnauthorized
+		apiName := info.FullMethod
+		index := strings.LastIndex(apiName, "/")
```

Comment:
Do we have to worry about the same method name being used in different paths? For instance

a/method1
b/method1

So now both of them "share" the same api name?

Reply:
Added `APIFullName` to `CallTarget` to make both cases convenient to handle. Let me know if you think this is overkill.

Reply:
TBH it seems like overkill. Might as well just include APIFullName, and we don't need the other field anymore.

Reply:
Removed APIFullName and assigned full name to APIName instead.